improver/blending/weighted_blend.py

```diff
@@ -612,103 +593,27 @@ def non_percentile_weights(self, cube, weights):
-
-        if weights:
-            weights_array = self.shape_weights(non_perc_slice, weights)
-            weights_array = iris.util.broadcast_to_shape(
-                weights_array, cube.shape, tuple(crd_dims)
-            )
-        else:
-            (number_of_fields,) = cube.coord(self.blend_coord).shape
-            weight = FLOAT_DTYPE(1.0 / number_of_fields)
-            weights_array = np.broadcast_to(weight, cube.shape)
-
-        (blend_dim,) = cube.coord_dims(self.blend_coord)
-        (perc_dim,) = cube.coord_dims(perc_coord)
-
-        # The percentile aggregator performs some coordinate reordering on
-        # the data. We don't have sufficient information in the aggregator
-        # to modify the weight order correctly, so we do it in advance.
-        weights_array = np.moveaxis(weights_array, (blend_dim, perc_dim), (0, 1))
-
-        # Check the weights add up to 1 across the blending dimension.
-        self.check_weights(weights_array, 0)
-
-        return weights_array
-
-    def percentile_weighted_mean(self, cube, weights, perc_coord):
+    def percentile_weighted_mean(self, cube, weights):
         """
         Blend percentile data using the weights provided.
 
         Args:
             cube (iris.cube.Cube):
-                The cube which is being blended over self.blend_coord.
+                The cube which is being blended over self.blend_coord. Assumes
+                self.blend_coord and percentile are leading coordinates (enforced
+                in process).
             weights (iris.cube.Cube):
                 Cube of blending weights.
-            perc_coord (iris.coords.DimCoord):
-                The percentile coordinate for this cube.
         Returns:
             iris.cube.Cube:
                 The cube with percentile values blended over self.blend_coord,
                 with suitable weightings applied.
         """
-        percentiles = np.array(perc_coord.points, dtype=FLOAT_DTYPE)
-        (perc_dim,) = cube.coord_dims(perc_coord.name())
+        non_perc_slice = next(cube.slices_over(PERC_COORD))
+        weights_array = self.get_weights_array(non_perc_slice, weights)
 
-        # The iris.analysis.Aggregator moves the coordinate being
-        # collapsed to index=-1 in initialisation, before the
-        # aggregation method is called. This reduces by 1 the index
-        # of all coordinates with an initial index higher than the
-        # collapsing coordinate. As we need to know the index of
-        # the percentile coordinate at a later step, if it will be
-        # changed by this process, we adjust our record (perc_dim)
-        # here.
-        if cube.coord_dims(self.blend_coord)[0] < perc_dim:
-            perc_dim -= 1
-
-        weights_array = self.percentile_weights(cube, weights, perc_coord)
+        # Check the weights add up to 1 across the blending dimension
+        self.check_weights(weights_array, 0)
```

Comment:
Check weights isn't needed for the non-percentile weights as np.average will always normalise the weights, I wonder if we should just always normalise the weights in the `get_weights_array` function and eliminate this altogether?

Reply:
`check_weights` doesn't actually do any normalisation - it just raises an error.  This should never happen.  So i am happy to remove this completely.

Reply:
OK, on discussion: I'm going to update the `check_weights` function into a `normalise_weights` function, which does the check and then normalises rather than throwing an error.  This is more useful to the user in the long term.